base/loading.jl

```diff
@@ -391,14 +435,20 @@ function _require(mod::Symbol)
                 warn(m, prefix="WARNING: ")
                 # fall-through, TODO: disable __precompile__(true) error so that the normal include will succeed
             else
-                return # success
+                M = m[end]
+                register_root_module(mod, M)
+                return M # success
             end
         end
 
         # just load the file normally via include
         # for unknown dependencies
+        local M
         try
-            Base.include_relative(Main, path)
+            # TODO: this is a hack to handle circular module references
+            tempmod = Module(temp_mod_name)
+            loaded_modules[mod] = tempmod
+            M = Base.include_relative(tempmod, path)
```

Comment:
This isn't currently expected to return a module. We'll need to add a deprecation warning here if we want to change the API.